### ISSUE TITLE
$$ properties passed on pageviews

### ISSUE THREAD

**Issue by troex** [angulartics/angulartics-segment#10]
My segment pageviews look like:

``` javascript
analytics.page('http://dev.local:4000/', {
  $$protocol: 'http',
  $$host: 'dev.local',
  $$port: 4000,
  $$path: '',
  $$search: {},
  $$hash: '',
  $$url: '',
  $$absUrl: 'http://dev.local:4000/',
  $$state: null,
  $$replace: false,
  name: 'http://dev.local:4000/',
  path: '/',
  referrer: 'http://dev.local:4000/pricing/',
  search: '',
  title: 'Site Title',
  url: 'http://dev.local:4000/'
});
```

I don't think `$$properies` should be passed to segment. Probably they are inherited from angulartics itself, shouldn't they be filtered?

**Comment by ProLoser** [comment]
`angular.copy()` filters these out. Perhaps you can submit a PR?

**Comment by aeldaly** [comment]
Doesn't look like angular.copy() filters them out

**Comment by ProLoser** [comment]
Sorry looks like `angular.toJson()` does it.

**Comment by aeldaly** [comment]
What do you think of this:

```
$analyticsProvider.registerPageTrack( function ( path, properties ) {
        try {
          properties = _.omit( properties, function ( value, key, object ) {
            return key.startsWith( '$$' );
          } );

          analytics.page( path, properties );
        } catch ( e ) {
          if ( !( e instanceof ReferenceError ) ) {
            throw e;
          }
        }
      } );
```

Only issue is it relies on underscore.

**Comment by ProLoser** [comment]
Why not use what's already built into angular?

**Comment by aeldaly** [comment]
So although angular.toJson does work, it's got issues :)

using _.omit:

```
analytics.page('/signin', {
  name: '/signin',
  referrer: 'http://localhost:9000/signin',
  title: 'title'
});
```

using toJson: 

```
analytics.page('/signin', '{}', {
  name: '{}',
  category: '/signin',
  path: '/signin',
  referrer: 'http://localhost:9000/signin',
  search: '',
  title: 'title',
  url: 'http://localhost:9000/signin'
});
```

**Comment by aeldaly** [comment]
k, this works like a charm:

```
$analyticsProvider.registerPageTrack( function ( path, properties ) {
        try {
          properties = JSON.parse(angular.toJson(properties));
          analytics.page( path, properties );
        } catch ( e ) {
          if ( !( e instanceof ReferenceError ) ) {
            throw e;
          }
        }
      } );
```

**Comment by aeldaly** [comment]
want me to create a PR? @ProLoser

**Comment by troex** [comment]
@aeldaly looks strange to me, maybe we should check how it's implemented for different providers?

**Comment by aeldaly** [comment]
@troex I've checked angulartics-newrelic-insights and angulartics-mixpanel and they both seem to just send properties as is. Maybe other providers filter $$ but segment does not?

**Comment by troex** [comment]
@aeldaly yeah strange indeed, I think `angular.toJson` is the right way to go. `name` being current page url in `page()` is normal for segment, so I don't see issue here

**Comment by aeldaly** [comment]
@troex the output was fixed when I added the `JSON.parse`.

How do I submit a PR? I'm not able to commit the branch I created. Do you want me to fork and submit a PR that way?

**Comment by ProLoser** [comment]
What is the original object you're working with? I don't understand why angular is yielding the response you described based on the data you were putting in.

**Comment by ProLoser** [comment]
```
console.log(angular.toJson({
  $$protocol: 'http',
  $$host: 'dev.local',
  $$port: 4000,
  $$path: '',
  $$search: {},
  $$hash: '',
  $$url: '',
  $$absUrl: 'http://dev.local:4000/',
  $$state: null,
  $$replace: false,
  name: 'http://dev.local:4000/',
  path: '/',
  referrer: 'http://dev.local:4000/pricing/',
  search: '',
  title: 'Site Title',
  url: 'http://dev.local:4000/'
}));

// yields:
{
  "name": "http://dev.local:4000/",
  "path": "/",
  "referrer": "http://dev.local:4000/pricing/",
  "search": "",
  "title": "Site Title",
  "url": "http://dev.local:4000/"
}
```

**Comment by aeldaly** [comment]
angular.toJson(properties) returns `'{}'` causing segment to generate this:

```
analytics.page('/signin', '{}', {
  name: '{}',
  category: '/signin',
  path: '/signin',
  referrer: 'http://localhost:9000/signin',
  search: '',
  title: 'title',
  url: 'http://localhost:9000/signin'
});
```

changing to `JSON.parse(angular.toJson(properties))` yields the expected result

**Comment by ProLoser** [comment]
What is the contents of `properties`?

**Comment by aeldaly** [comment]
@ProLoser did you look at the result using the debugger in segment.io?

**Comment by aeldaly** [comment]
Wondering if there's a more efficient way to have this discussion. Screenhero maybe?

**Comment by ProLoser** [comment]
I want to understand why you're getting an object as the value of the `name` property. I dropped in [the object you pasted above](https://github.com/angulartics/angulartics-segment/issues/10#issue-151955656). If you're saying the properties object contains something different than what you posted above, such as having an object as it's value, that's a symptom of a far larger problem in my opinion and I'd rather not go about fixing symptoms with hacks vs identifying the problem.

I no longer use angular whatsoever so most of my assistance is just trying to be helpful, I can't really screenshare with people at this time.

**Comment by ProLoser** [comment]
I also think this discussion should probably be relocated to the core library as this is not isolated to segment

**Comment by troex** [comment]
@aeldaly I agree with @ProLoser, how do you call `.page()` method in your app? I think it's something messing in your application code and not `angulartics-segment` itself. I'll have time later today to review your solution in my app, I'll write back my results.

**Comment by aeldaly** [comment]
I'm using this:

`$analytics.eventTrack( 'Logged In' );`

this code:

```
      $analyticsProvider.registerPageTrack( function ( path, properties ) {
        try {
          console.log(properties);
          properties = JSON.parse(angular.toJson(properties));
          console.log(properties);
          analytics.page( path, properties );
        } catch ( e ) {
          if ( !( e instanceof ReferenceError ) ) {
            throw e;
          }
        }
      } );
```

produces:
in console:

```
angulartics-segment.js:21 LocationHtml5Url {$$html5: true, $$protocol: "http", $$host: "localhost", $$port: 9000, $$path: "/welcome"…}$$absUrl: "http://localhost:9000/welcome"$$compose: ()$$hash: ""$$host: "localhost"$$html5: true$$parse: (url)$$parseLinkUrl: (url, relHref)$$path: "/welcome"$$port: 9000$$protocol: "http"$$replace: false$$search: Object$$state: null$$url: "/welcome"__proto__: Object
angulartics-segment.js:23 Object {}
```

in segment:

```
analytics.page('/signin', {
  name: '/signin',
  path: '/signin',
  referrer: 'http://localhost:9000/signin',
  search: '',
  title: 'title',
  url: 'http://localhost:9000/signin'
});
```

**Comment by ProLoser** [comment]
@aeldaly you're firing an `eventTrack` and your callback function is a `pageTrack` so I'm now doubly confused:
1. Why are you looking at pageTracking code for eventTracking?  
2. What is the contents of your `properties` object (before any parsing)

**Comment by ProLoser** [comment]
Okay to be clear, `$analytics.eventTrack( 'Logged In' );` has nothing to do with any of this.

Secondly, I think the `properties` object is actually the `$location` object which probably has some weird data inside.

See this line: https://github.com/angulartics/angulartics/blob/master/src/angulartics.js#L190

I have no idea what `$location` puts into the `name` property or why. This is also some slight area of complication because

**Comment by troex** [comment]
Okay I did some research on segment docs and took a look at other libs for angular + segment that I used and committed to two years ago. The actual problem comes from that `properties` are passed to segment - `analytics.page(path, properties);` in fact we only need to pass `analytics.page(path);`

All the rest properties will be automatically populated by segmentio script, injecting $location (in any way with `angular.toJson`, `JSON.parse` or raw) does not give any benefits instead it messes up segment stats. The only reasonable property that we might pass is `referrer` with previous state url, but since `angulartics` doesn't provide it there is really NO need to pass properties at all.

**Comment by troex** [comment]
By default segment script adds next properties: `path, referrer, search, title, url` if for some reason you think one of these values is incorrect, like `url` being `http://mysite.com/#!/state/action` instead of `http://mysite.com/state/action` - you should tweak it manually.

In my case it's not a problem since I use angular for stateless app (I don't have state changes or any kind of routers) but may be the case for more common angular app.

Hope this clarifies the problem.

**Comment by troex** [comment]
@aeldaly can you please remove `properties` from `analytics.page(path, properties);` and see if it works correctly for you?

**Comment by timelf123** [comment]
Released 0.1.5 with this https://github.com/angulartics/angulartics-segment/issues/12
